packages/vite/src/node/plugins/html.ts

```diff
@@ -1265,6 +1265,45 @@ export function resolveHtmlTransforms(
   return [preHooks, normalHooks, postHooks]
 }
 
+// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/head#see_also
+const HeadElement: Record<string, boolean> = [
+  'title',
+  'base',
+  'link',
+  'style',
+  'meta',
+  'script',
+  'noscript',
+  'template',
+].reduce(
+  (res, tag) => ({
+    ...res,
+    [tag]: true,
+  }),
+  {},
+)
+
+function headTagInsertCheck(
+  tags: HtmlTagDescriptor[],
+  ctx: IndexHtmlTransformContext,
+) {
+  if (!tags.length) return
+  const { logger } = ctx.server?.config || {}
+  const message = tags.reduce<string[]>((res, tagDescriptor) => {
+    if (!HeadElement[tagDescriptor.tag]) res.push(`<${tagDescriptor.tag}>`)
```

Comment:
Note for other reviewers: I confirmed that custom elements are not allowed in `<head>`.